presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java

```diff
@@ -764,34 +780,100 @@ public Page getNextPage()
-                if (reorderFilters && (currentRowGroup & 0x3) != 0 && currentRowGroup != 0) {
-                    // Reconsider filter order every 4 row groups.
-                    reader.maybeReorderFilters();
+                if (currentRowGroup != 0 && (currentRowGroup % ROW_GROUP_REVIEW_INTERVAL == 0)) {
+                    // Decay row size stats and reconsider filter
+                    // order every ROW_GROUP_REVIEW_INTERVAL row
+                    // groups.
+                    numAriaRows /= 2;
+                    numAriaBytes /= 2;
+                    if (reorderFilters) {
+                        reader.maybeReorderFilters();
+                    }
+                }
                 }
+            if (qualifyingSet.isEmpty()) {
+                qualifyingSet.setRange(qualifyingSet.getEnd(), Math.min(qualifyingSet.getEnd() + ariaBatchRows, currentGroupRowCount));
             }
-            qualifyingSet.clearTruncationPosition();
-            reader.advance();
+            int bytesBeforeAdvance = reader.getResultSizeInBytes();
+            int numResultsBeforeAdvance = reader.getNumResults();
+            try {
+                reader.advance();
+            }
+            catch (BatchTooLargeException e) {
+                // The reader ran out of budget. Retry with a smaller
+                // input qualifying set. First remove any values added
+                // by the last advance.  The input qualifying set is
+                // unaltered when the reader throws this exception.
+                // If we are at minimum batch size we must also have unlimited budget. Check that this path is not taken with minimum batch size.
+                verify(qualifyingSet.getEnd() - qualifyingSet.getPositions()[0] > MIN_BATCH_ROWS, "Running out of budget with minimum batch size");
+                reader.compactValues(new int[0], numResultsBeforeAdvance, 0);
+                ariaBatchRows = Math.max(MIN_BATCH_ROWS, ariaBatchRows / BATCH_HARD_SHRINK_FACTOR);
+                int[] rows = qualifyingSet.getPositions();
+                int numRows = Math.min(ariaBatchRows, qualifyingSet.getEnd() - rows[0]);
+                qualifyingSet.setPositionCount(numRows);
+                // The first row not in scope is 1 above the last in scope.
+                qualifyingSet.setEnd(rows[numRows - 1] + 1);
+                setReadersToCurrentRowGroup();
+                if (reader.getNumResults() > 0) {
+                    setReaderBudget();
+                    return resultPage();
+                }
+                continue;
+            }
+            long rowSize = numAriaBytes / (1 + numAriaRows);
+            int numHitsInBatch = toIntExact(targetResultBytes / (1 + rowSize));
+
             numResults = reader.getNumResults();
-            if (numResults > targetResultRows || reader.getResultSizeInBytes() > targetResultBytes * 8 / 10) {
+            long readerBytes = reader.getResultSizeInBytes();
```

Comment:
- This method is very long and dense. Consider, extracting helper method for adjusting numAriaRows. 
- rowSize and numHitsInBatch are used only in one branch below. Let's move them there and rename to averageBytesPerRow and averageRowsInBatch.
- What's the purpose of targetResultRows? It seems unnecessary.

```
            numResults = reader.getNumResults();
            adjustBatchSize(bytesBeforeAdvance, numResultsBeforeAdvance);
            setReaderBudget();
            if (numResults > targetResultRows || reader.getResultSizeInBytes() > targetResultBytes) {
                return resultPage();
            }
...
    private void adjustBatchSize(long bytesBeforeAdvance, int numResultsBeforeAdvance)
    {
        long readerBytes = reader.getResultSizeInBytes();
        long increment = readerBytes - bytesBeforeAdvance;
        numAriaBytes += increment;
        numAriaRows += numResults - numResultsBeforeAdvance;
        if (increment > targetResultBytes && ariaBatchRows > MIN_BATCH_ROWS) {
            // Make batch smaller if not already at minimum.
            ariaBatchRows = Math.max(ariaBatchRows / BATCH_SOFT_SHRINK_FACTOR, MIN_BATCH_ROWS);
        }
        else if (increment == 0 && numAriaRows > 0) {
            // If skipping over no hits, do not make the batch
            // size too much larger than the number of hits that
            // fit in a batch so as to avoid a retry.
            long averageBytesPerRow = numAriaBytes / (1 + numAriaRows);
            int averageRowsInBatch = toIntExact(targetResultBytes / (1 + averageBytesPerRow));
            ariaBatchRows = Math.min(averageRowsInBatch * 2, ariaBatchRows * BATCH_SOFT_SHRINK_FACTOR);
        }
        else if (increment < targetResultBytes / 8) {
            ariaBatchRows = Math.min(ariaBatchRows * BATCH_GROW_FACTOR, currentGroupRowCount);
        }
    }
```